### PR TITLE
Rancher Dashboard Integration

### PR DESCRIPTION
**Pull request by rak-phillip**
This adds a simplified version of Rancher Dashboard to Rancher Desktop by making use of embedded versions of the Steve API and specialized builds of Rancher Dashboard.

### CODE REVIEW

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Didn't work on macos -- it complained that it couldn't find `steve` -- I had built a macos version of it and placed it in `resources/darwin/bin`.

**Comment by rak-phillip** [comment]
> Didn't work on macos -- it complained that it couldn't find `steve` -- I had built a macos version of it and placed it in `resources/darwin/bin`.

Issue should be fixed, thanks for taking a look 🙂

**Review by ericpromislow — CHANGES_REQUESTED** [review]
I think the dashboard needs to automatically refresh. I brought it up, then created a namespace and created a pod in that namespace, and even a minute after the pod was running, none of the new resources showed up in the dashboard. When I closed the dashboard and then reopened it, it showed them.

**Comment by rak-phillip** [comment]
> I think the dashboard needs to automatically refresh. I brought it up, then created a namespace and created a pod in that namespace, and even a minute after the pod was running, none of the new resources showed up in the dashboard. When I closed the dashboard and then reopened it, it showed them.

This is confirmed an issue with dashboard and will need to be patched on that end.

**Comment by rak-phillip** [comment]
In reviewing my changes, I noted that consideration needs to be made based on Kubernetes being enabled or not. Moving back to draft

**Comment by rak-phillip** [comment]
@ericpromislow @mook-as looking for feedback on this one, thanks 🙂

**Comment by ericpromislow** [comment]
macos Big Sur: First problem on current run shows up in the main console output:
```
Spawned child pid: 69613
stdout: Incorrect Usage. flag provided but not defined: -offline

stdout: NAME:
   steve

USAGE:
   steve [global options] command [command options] [arguments...]
[rest of usage omitted]…
```

and the Rancher Dashboard shows up as a blank screen.

**Comment by ericpromislow** [comment]
On the positive side, when k8s is disabled the Dashboard menu item is correctly disabled.

**Comment by rak-phillip** [comment]
> macos Big Sur: First problem on current run shows up in the main console output:
> 
> ```
> Spawned child pid: 69613
> stdout: Incorrect Usage. flag provided but not defined: -offline
> 
> stdout: NAME:
>    steve
> 
> USAGE:
>    steve [global options] command [command options] [arguments...]
> [rest of usage omitted]…
> ```
> 
> and the Rancher Dashboard shows up as a blank screen.

The error makes me think that you have an older version of the Steve binary in your resources from when you last reviewed. You might need to delete Steve and Rancher Dashboard from resources and `npm install` again to get the latest version.

**Comment by ericpromislow** [comment]
I ran `git clean -ffd ; npm run postinstall` initially, but I'll try a more draconian cleanup.

**Comment by ericpromislow** [comment]
Yes, a full cleaning fixed that problem. Lots of logs going to the main console -- can you redirect them to a `.../logs/` , or are they coming from dashboard itself?

In the UI, I initially saw the top of the window, but now it's scrolled off the top of the window and I can't get it back -- so I can't change the namespaces.

img-1270a

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Scrolling problems in the UI -- when I go into Workload/Deployments I lose the top part of the screen and can't scroll into it.

**Comment by rak-phillip** [comment]
> Yes, a full cleaning fixed that problem. Lots of logs going to the main console -- can you redirect them to a `.../logs/` , or are they coming from dashboard itself?
> 
> In the UI, I initially saw the top of the window, but now it's scrolled off the top of the window and I can't get it back -- so I can't change the namespaces.
> 
> img-1270a

Should be fixed, thanks 🙂 

It looks like the issue is with resizing the window on macOS. `win.getContentSize()` is the method I should have been using to resize consistently across platforms..

**Comment by rak-phillip** [comment]
@ericpromislow also redirected Steve console output to a log file

File: src/pages/K8s.vue [hunk 1]
```diff
@@ -361,12 +361,15 @@ export default {
...
-          '\n\nDo you want to proceed?'].join('');
+        const confirmationMessage = [
+          `${ value ? 'Enabling' : 'Disabling' } Kubernetes requires a restart. `,
+          '\n\nDo you want to proceed?'
+        ].join('');
 
         if (confirm(confirmationMessage)) {
           try {
             await ipcRenderer.invoke('settings-write', { kubernetes: { enabled: value } });
+            ipcRenderer.send('dashboard-close');
```
Comment (ericpromislow) [inline src/pages/K8s.vue:372]: Thinking about the CLI, the request to shutdown the dashboard should be sent from the code in `background.ts` that processes the `settings-write` event. Thinking along the lines of having UI event handlers do as little as possible, which in most cases should be working with widgets and sending events.

File: src/k8s-engine/steve.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+const console = Logging.steve;
+
+/**
+ * @description Singleton that manages the lifecycle of the Steve API
+ */
+export class Steve {
+  private static instance: Steve;
+  private process!: ChildProcess;
+  private uiPath() {
+    return path.join(process.cwd(), 'resources', 'rancher-dashboard');
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/k8s-engine/steve.ts]: Thanks for the suggestion, making use of `resources.get()` definitely cleans things up considerable here 🙂

File: src/pages/K8s.vue (hunk 1 above)
Comment (rak-phillip) [reply src/pages/K8s.vue]: I removed this and implemented some logic in the `settings-write` listener to check if kubernetes is disabled before closing the dashboard.

File: src/main/networking/index.ts [hunk 3]
```diff
@@ -35,7 +35,8 @@ export default function setupNetworking() {
 
   // Set up certificate handling for system certificates on Windows and macOS
   Electron.app.on('certificate-error', async(event, webContents, url, error, certificate, callback) => {
-    const dashboardUrls = ['https://127.0.0.1:9443', 'wss://127.0.0.1:9443'];
+    const tlsPort = 9443;
+    const dashboardUrls = [`https://127.0.0.1:${ tlsPort }`, `wss://127.0.0.1: ${ tlsPort }`];
```
Comment (ericpromislow) [inline src/main/networking/index.ts:39]: Shouldn't be a space after the wss ':'

File: src/window/dashboard.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  view.setBounds({
+    x:      0,
+    y:      0,
+    width:  windowSize[0],
+    height: windowSize[1],
+  });
+
+  view.setAutoResize({ width: true, height: true });
+
+  view.webContents.loadURL(dashboardURL);
```
Comment (ericpromislow) [inline src/window/dashboard.ts:27]: I just saw that this returns an unhandled promise. Best to just catch any errors like:
```
view.webContents.loadURL(dashboardURL).catch((err) => {
  console.log(`Can't load the dashboard URL ${ dashboardURL }: `, err);
}

**Review by ericpromislow — CHANGES_REQUESTED** [review]

File: src/main/networking/index.ts (hunk 3 above)
Comment (rak-phillip) [reply src/main/networking/index.ts]: Good eye - fixed

**Review by ericpromislow — APPROVED** [review]

**Pull request merged**